Add angle bound support to target constraints

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a new capability to the target for representing angle bounds on a gate in the target and a transpiler pass to enforce the bounds are respected. To enable this functionality it requires that the target constructor provides a function to do the angle wrapping/folding so that the transpiler knows how to do the transformation of arbitrary gate appropriately. Since the transpiler can't make any assumption about the bounds or the angles in use, or the gates that will have bounds applied. So coming up with a general transpiler pass is not possible, however a user providing the conversion function when applying a bound makes it possible for the preset pass managers to reason about gates with angle bounds and work with them.

### Details and comments

Fixes #13784

TODO:
- [x] Fix lint
- [x] Fix bug in preset pass managers leading to out of bounds gates and/or fix test code

mtreinish committed:
Add angle bound support to target constraints

This commit adds a new capability to the target for representing angle
bounds on a gate in the target and a transpiler pass to enforce the
bounds are respected. To enable this functionality it requires that the
target constructor provides a function to do the angle wrapping/folding
so that the transpiler knows how to do the transformation of arbitrary
gate appropriately. Since the transpiler can't make any assumption about
the bounds or the angles in use, or the gates that will have bounds
applied. So coming up with a general transpiler pass is not possible,
however a user providing the conversion function when applying a bound
makes it possible for the preset pass maangers to reason about gates
with angle bounds and work with them.

This new feature is strictly additive to the Target which keeps it
backwards compatible. It does mean that for existing workflows a user
will need to opt-in to checking the angle bounds.

Fixes #13784

mtreinish committed:
Fix lint

mtreinish committed:
Add missing pickle of angle bounds

mtreinish committed:
Fix fold rzz function and fix failing test

mtreinish committed:
Add in bounds angle to workflow test

mtreinish commented:
I think this should be ready for review now. I have the full path testing working now (it was all bugs in the rzz folding function). The `WrapAngles` pass might need a bit more direct testing to cover all the cases. I'll look at that in parallel, but we should have sufficient coverage through what's there and the full path transpiler tests at this point.

mtreinish committed:
Merge branch 'main' into wrap-angles-constraints

mtreinish committed:
Missing imports from release note example

mtreinish committed:
Handle no-target in translation generation

mtreinish committed:
Fix release note example

jakelishman commented on .gitignore:
```diff
@@ -153,4 +153,4 @@ test/visual/mpl/visual_test_failures/*.png
 # Added by cargo
 
-/target
+./target
```
This line was already correct, the problem is the `target/` on line 81.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14406/commits/1e6c8f71e0a75c61c9817f484ca0afd1b2d296e2

jakelishman commented on crates/pyext/src/lib.rs:
```diff
@@ -74,5 +74,6 @@ fn _accelerate(m: &Bound<PyModule>) -> PyResult<()> {
     add_submodule(m, ::qiskit_circuit::converters::converters, "converters")?;
     add_submodule(m, ::qiskit_qasm2::qasm2, "qasm2")?;
     add_submodule(m, ::qiskit_qasm3::qasm3, "qasm3")?;
+    add_submodule(m, ::qiskit_transpiler::passes::wrap_angles_mod, "wrap_angles")?;
```
Once upon a time, this list was sorted...

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -885,8 +928,65 @@ impl Target {
+            .map_err(|err| TranspilerError::new_err(err.to_string()))?;
+        self.angle_bounds.insert(name, new_bound);
         Ok(())
     }
```
Spiritually, I don't think this should be separate to `add_instruction`.  The angle bound is part of the support, and mutation of the `Target` is supposed to be additive only, whereas this adds a restriction later.  I think the signature of `add_instruction` should be
```python
def add_instruction(
    self,
    obj: Instruction,
    properties: dict[Qargs, InstructionProperties] = None,
    name: str = None,
    bounds: Optional[tuple[Bound, ...]] = None,
): ...
```
or something, so the bounds are specified at the same time.

mtreinish replied:
I agree, I moved this into `add_instruction` in https://github.com/Qiskit/qiskit/pull/14406/commits/f0e18c0c88551f44f5210b7132acc14a3f5b873e

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -885,8 +928,65 @@ impl Target {
+    /// Does the gate have an angle bound?
+    pub fn gate_has_angle_bound(&self, name: &str) -> bool {
+        self.angle_bounds.contains_key(name)
+    }
```
Super minor: a gate can still have more than one angle, and it's got an upper and lower bound anyway, so I think it'd still be best as `gate_has_angle_bounds`?

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14406/commits/2763c0db5c4ca7d72ceff4f86bd95f179de13568

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -831,6 +861,19 @@ impl Target {
+                }
+            };
+            bounds_dict.set_item(gate, (bounds.bounds().to_vec(), callback))?;
+        }
```
We will need to consider a story around parallelisation in the future here (goes for the Python form too), but I don't mind leaving it for now.

mtreinish replied:
Realistically if we're relying on pickle for IPC in a Python multiprocessing context I don't think there is a realistic option here if something is defined via a function pointer in C or Rust there isn't much we'll be able to do to pass that to a child process and have it work.

My original thinking around the `Native` variant was to use it as a place for vendors (IBM and AQT being the examples I know of at the moment) to contribute default rules to the global registry to potentially offer an accelerate version. In that model the multiprocessing story would be always add the `Native` variants in `__new__` and drop them from the pickle output. But this obviously doesn't work because the rules are based on the target not the abstract gate definition so we can't use this model.

The only real use case I can think of is this is for C extensions in the future and we're already going to be limited in the transpiler for C that it only works in a standalone context and can't be mixed with Python. If you're in a Python context you need to use the Python transpiler. If we document it very explicitly that in a mixed Python/C context you can only use the transpiler from C then I think we should be safe here because you'd only be able to define native callbacks in a context where we can't ever be using multiprocessing.

jakelishman commented on crates/transpiler/src/target/bounds.rs:
```diff
@@ -0,0 +1,111 @@
+pub(crate) enum CallbackType {
+    Python(PyObject),
+    Native(fn(&[f64]) -> DAGCircuit),
+}
```
I got into it a lot more in the top PR comment, but I don't think this signature is right in general.  Python can have dynamic closures so the cost is lessened, but even within that I have concerns.

mtreinish replied:
I updated the signature here to take qubit indices as a second argument. That should provide enough context for the caller to figure this out since they need to register the callback themselves. That being said I was debating adding a target reference here too just so the calback can query the target directly, I still think we probably should add it and can if you agree.

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -511,12 +513,22 @@ impl Target {
+    ///         when building custom transpilation pipelines you will want to set this to ``True``.
+    ///         If any of the values in ``parameters`` are set to be :class:`.ParameterExpression`
+    ///         instances this flag will have no effect as angle bounds only impact
+    ///         non-parameterized operations in the circuit.
```
I got into this a lot more in the main review comment, but flagging here: I think the addition of this method defaulting to `false` fundamentally changes `instruction_supported`.  If anything, I think it should default to `true`, and if we ever use it in Qiskit code, we should adjust our code.  We ought to cause passes to fail loudly, not silently produce incorrect output - I think any claim of "backwards compatibility" already comes with a lot of qualifiers, and that this errs on the wrong side in the name of something we're already breaking.

For similar reasons, I think the Rust-space `Target::instruction_supported` should also check the bound angles; I think it's a foot-gun waiting to go off if we don't deal with it now, but silently soften the method to returning "instruction is supported if you do some more translation work".

mtreinish replied:
I flipped the default in https://github.com/Qiskit/qiskit/pull/14406/commits/240dab7a24c52fc877c44aa9789418024dad712f 

The rust space method doesn't seem to understand parameters at all!? I think we can leave this for a follow up potentially.

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -885,8 +928,65 @@ impl Target {
+    ///         This function is passed a float value for the angle and returns a :class:`.DAGCircuit`
+    ///         representing the equivalent circuit for ``gate`` with that angle value. It is expected
+    ///         that the callback will generate a gate in terms of other supported instructions on the
+    ///         target. This function will be called by the :clas:`.WrapAngles`
```
The argument isn't called `wrap_function` - it's called `callback`.

> It is expected that the callback will generate a gate in terms of other supported instructions on the target.

The callback function doesn't get enough information to fulfill this requirement.  Consider the heterogeneous QPU with `[sx, rz]` on 0 and 1, `[s, rx]` on 2 and 3, and a bound-angle `rzz` on pairs `(0, 1)` and `(2, 3)`.  A callback function `fn(&[f64]) -> DAGCircuit` cannot reliably produce a circuit that has supported instructions.

mtreinish replied:
Resolved with: https://github.com/Qiskit/qiskit/pull/14406/commits/f0e18c0c88551f44f5210b7132acc14a3f5b873e

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1409,6 +1509,71 @@ impl Target {
+            Some(bounds) => Some(bounds.get_replacement_circuit(params)).transpose(),
+            None => Ok(None),
+        }
+    }
```
This really doesn't feel right for a `Target` method.

mtreinish replied:
It is no longer after https://github.com/Qiskit/qiskit/pull/14406/commits/f0e18c0c88551f44f5210b7132acc14a3f5b873e

jakelishman reviewed (commented):
My main thoughts are: adding bounds is fine, but I'm pretty suspicious of adding this "callback" transformation logic to the `Target`, and I'm concerned that we don't have enough time before 2.1 to think through the implications of it.

I don't agree with two commit-message claims:

> This new feature is strictly additive to the Target which keeps it backwards compatible. It does mean that for existing workflows a user will need to opt-in to checking the angle bounds.

I get into this more below, but the summary is that it's IMO only backwards compatible for use cases that look like "user calling a preset passmanager with built-in plugins".  I think we need to add the bounds anyway, but I don't think we just make API decisions around making the `Target` backwards compatible, because it isn't - `instruction_supported` no longer firmly means "instruction supported", which could cause downstream passes to silently break their own contracts (e.g. the existing `BasisConstructor` translation plugin is no longer a valid translation plugin after this change).  We can make the "preset passmanager" use case backwards compatible in many different ways.

> coming up with a general transpiler pass is not possible, however a user providing the conversion function when applying a bound makes it possible for the preset pass managers to reason about gates with angle bounds and work with them.

I don't think this justifies that the conversion must be in the `Target`.  Nothing compels you on `Target` creation to have ensured full coverage of `EquivalenceLibrary` rules that let the `BasisTranslator` work, so we have a prior on making non-realisible translation targets, and nothing compels users to _call_ the `WrapAngles` pass, so enforcing that the bound-conversion logic sits inside the `Target` doesn't actually mean that it gets applied.

---

In no particular order, and I'm not saying all of these are hard blockers, and in some cases maybe we can just dismiss the concern; I'm just trying to think through implications:

1. We don't specify the basis-translation rules in the `Target`, for example, and this means that Qiskit can't easily have built-in high-performance rewriters for our standard gates, and it isn't offering an easy path to extensibility.  It feels like an `Instruction.definition` vs `BasisTranslator` kind of issue, maybe: we require a "blessed" version to exist without any context, and we're relying on the creator of the `Target` to put any context they need into the conversion function at `Target`-creation time (and precluding context other than the other supported instructions in the global sense).
2. Adding the concept of angle bounds to me means we ought to consider how this scales if different qargs have different angle bounds.  Certainly it's representable, but I'm concerned that the static function-pointer callback will require a lot of hard-coding in languages without dynamic function definitions (Python).  For example, you can write the RZZ rewriter pretty generally to hit any $pi/2$-width interval, but to use the `Native` option here, you'd have to have a hard-coded entry point in your compiled binary for each individual range you might care about, which precludes being fed it dynamically from a calibration run.  The general concept of "this doesn't allow closures" is my issue; we'd need something like `fn(&[f64], *mut void) -> DAGCircuit` and to store a type-erased pointer to a "state" object to permit this sort of dynamic specification in compiled languages.

---

Here's a sketch of a possible alternative, which I also haven't thought through and I'm not suggesting a pivot to this - I just want an alternative to framework a comparative discussion, since this is really core data model stuff for the compiler.

Let's say we have a registry, like the `BasisTranslator`'s global `EquivalenceLibrary`, which stores angle rewriters, or perhaps angle-rewriter _factories_ (functions / vtables).  Effectively, it's a mapping of `str: list[RewriterFactory]` or something - the keys are the same names as in the `Target`, and we can potentially have more than one factory (if we want to open the door to having better/worse performing ones and we somehow select).  I guess the something like:
```rust
type rewriterfunc = fn(&[f64], *const void) -> DAGCircuit;
type freestatefunc = fn(*const void);
type factoryfunc = fn(&Context) -> (rewriterfunc, freestatefunc, *const void);
```
where:
- `rewriterfunc` is the same as the existing `Native` callback, except it accepts an arbitrary state object too, so it can be dynamic.
- `freestatefunc` is a function pointer to something that can free the state object.
- `factoryfunc` is a function that takes some `Context` about the rewriting, and returns a suitable rewriter (if any), a type-erased state object for the rewriter, and a method to release the ownership of the state object.
- `Context` is some object that represents the context of the rewriting. That would include some subset of: the name of the object, the angle bounds, potentially a pointer to the `Target` and the qubits we're acting on?  There's a choice to be made here about how much complexity we expose, and how much should just be implemented in custom peephole optimisation passes.

I think the concern of factory-vs-direct rewriter is the same between both the PR and this structure, so that's just for illustration.

Mostly, I want to think about whether these rewriters should be in the `Target`, or if they should come in through some alternative registry that the `WrapAngles` pass uses.  If it's possible to add them to the `Target`, it's possible to add them to an external registry too, but an external registry can include fields that are pre-populated by Qiskit, so we can have built-in methods, and don't require hardware vendors to reinvent the wheel for things we can centralise in a general manner.

---

I _think_ (correct me if wrong) that a large part of the reason to add the bound-resolving callbacks to the `Target` was to be able to claim backwards compatibility in the default call to `instruction_supported`, with the argument being "even if it violates the angle bounds it still _can be_ supported because you can use that callback".

If so: I don't think that's enough justification to say the callback has to be on the `Target`; I think we're breaking backwards compatibility anyway.  _At least_, that bakes in several further assumptions that are not satisfied by the `Target` alone:
1. that somebody _does_ call that `WrapAngles` pass at some point.  For example, this is a problem because `GatesInBasis` was supposed to be the pass that told you if the circuit was currently properly compiled, but this change breaks that implication.  (Fwiw, I would have had less of a problem with this before `BackendV2` when there was a responsibility split between `CheckMap` and `GatesInBasis`, but now they both take the `Target`, `GatesInBasis` to me implicitly has a stronger contract.)  Even if we accept the break in `GatesInBasis`, it breaks an assumption of the method from any user pass; `instruction_supported` no longer means "instruction supported", it means "instruction can be supported if you do some additional steps".
2. that the rewriter is infallible.  This is probably something we _can_ require, but at the moment I don't think there's any clarity on what gates the rewriter is allowed to include in its return value.  Is it limited to one invocation of the concerned gate?  Is it limited that any additional added gates are also hardware-compatible for the qubit?  If so (and I believe that that _is_ the intention right now): what if the `Target` is heterogeneous and the other supported instructions might be different for different `qargs` that this gate is valid on (this speaks to the `Context` I was using above in my examples).

I think in the specific context of "you're only using the built-in plugins for the Qiskit preset passmanagers", the end result is that as a user you wouldn't notice that the angle bounds have been applied.  But it causes custom translation stages to silently break their contracts through no fault of their own, for example, and I'd call that a backwards-compatibility break.

To be clear: I don't think we can do this without breaking backwards compatibility in some way, and I still think we should add the bounds.  I don't think there's a way around breaking custom translation stages (well, there _is_, but it involves grafting on a versioning mechanism to the pass-manager stage plugin interface). But I think recognising that "backwards compatibility" has big qualifications on it anyway means that we have more scope for exploring other avenues.

---

Last top-level comment: I think there should be more documentation / examples in the main `Target` docstring about angle bounds.  Imo the main discussion of them should be in the description of its data model, not in the function that adds an entry in the data model.

jakelishman commented:
Sorry that comment was really long - I didn't have a super clear picture in my head of all my concerns, so it's hard to be concise.  Something feels quite wrong to me about putting the conversion logic in the `Target`, so if that's the path we do end up taking, I want to really make sure we've examined a bunch of the implications of it, justified what "backwards compatibility" means in this context, and have a strong argument for why it's in the `Target` and not in a user-extensible / Qiskit-extensible registry associated with the `WrapAngles` pass.

mtreinish committed:
Merge branch 'main' into wrap-angles-constraints

mtreinish committed:
Merge remote-tracking branch 'origin/main' into wrap-angles-constraints

mtreinish committed:
Fix gitignore

mtreinish committed:
Pivot to standalone registry

This commit pivots the user interface to the wrapping callbacks. Instead
of putting it all in the target this adds a separate data structure
analgous to the equivalence library for the basis translator but for
WrapAngles. This makes the target solely concerned with the QPU
constraints (which is it's purpose) and the backend provider is
responsible for populating the global registry instance to make the pass
work.

mtreinish committed:
Change default on check_angle_bounds param

mtreinish committed:
Rename gate_has_angle_bound -> gate_has_angle_bounds

mtreinish committed:
Merge branch 'main' into wrap-angles-constraints

mtreinish commented:
> My main thoughts are: adding bounds is fine, but I'm pretty suspicious of adding this "callback" transformation logic to the Target, and I'm concerned that we don't have enough time before 2.1 to think through the implications of it. ...

I've reworked the PR to use a registry class `WrapAngleRegistry` (the names of everything are terrible in this PR feel free to propose alternatives) for the `WrapAngles` pass. There is a global instance that can be added to for use with the preset pass managers. I removed the callback mechanism from the `Target` and it only exists in the registry class now.

mtreinish commented on crates/transpiler/src/angle_bound_registry.rs:
```diff
@@ -0,0 +1,157 @@
+                        name,
+                        *capsule.reference::<fn(&[f64], &[PhysicalQubit]) -> DAGCircuit>(),
+                    );
+                }
```
Oops, I forgot to remove this before pushing it up. I need to remove this and replace it with the exception that was previously in here. I was just playing around to see if I could get this to compile in Rust. It isn't safe and couldn't actually work because this is just wrapping a Rust function pointer in a `PyCapsule` and passing it to pickle/dill (`dill` claims to support serializing `PyCapsule` but I don't know what it's doing). We aren't going to be able to call a function pointer across a process boundary (which is why pickle support is needed here) so erroring is all we can do.

mtreinish committed:
Fix lint

mtreinish committed:
Remove unsound PyCapsule usage

Returning a PyCapsule of the function pointer from Rust to pickle isn't
going to do anything useful since the pickle is typically used for IPC
and the pointer won't be valid in a separate process. This removes that
and just raises an error.

mtreinish committed:
Add missing import to release note example

mtreinish committed:
Merge remote-tracking branch 'origin/main' into wrap-angles-constraints

mtreinish committed:
Improve docs

This commit expands the target side documentation for angle bounds
adding an overview of the data model to the class docstring and adding
missing details to the method docstrings.

mtreinish committed:
Fix release note example

mtreinish committed:
Improve WrapAngles pass documentation

mtreinish committed:
Fix docs build failures

mtreinish committed:
Merge remote-tracking branch 'origin/main' into wrap-angles-constraints

mtreinish committed:
Merge remote-tracking branch 'origin/main' into wrap-angles-constraints

kt474 commented on qiskit/transpiler/target.py:
```diff
@@ -159,6 +159,38 @@ class Target(BaseTarget):
+    You can check if any operations in the target have angle bounds set with,
+    :meth:`.has_angle_bounds` and also if a specific name in the target has
+    angle bounds set with :meth`.gate_has_angle_bounds`. Whether a paricular
+    set of parameter values confroms to the angle bounds can be checked
```
```suggestion
    angle bounds set with :meth`.gate_has_angle_bounds`. Whether a particular
    set of parameter values conforms to the angle bounds can be checked
```

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

kt474 commented on releasenotes/notes/add-wrap-instructions-32335ea783203be9.yaml:
```diff
@@ -0,0 +1,110 @@
+    an inclusive bound. For example, you can add ``RZGate(math.pi)`` to a :class:`.Target` and restrict
+    the angle value between the values 0 and :math:`2\pi`.
+
+    There are several methods available for working with the angle bouds on the target. The first is
```
```suggestion
    There are several methods available for working with the angle bounds on the target. The first is
```

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

kt474 commented on qiskit/transpiler/passes/utils/wrap_angles.py:
```diff
@@ -0,0 +1,88 @@
+    the target will be called to substitute the gate outside the bound with an equivalent subcircuit.
+    This pass does not run on gates that are parameterized, even if the gate has unparameterized
+    parameters outside a specified bound. If there are parameterized gates in the circuit they will
+    be ignored by this pass as bound angles are neceessary to transform the gate. For example:
```
```suggestion
    be ignored by this pass as bound angles are necessary to transform the gate. For example:
```

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

kt474 commented on releasenotes/notes/add-wrap-instructions-32335ea783203be9.yaml:
```diff
@@ -0,0 +1,110 @@
+    Added a new transpiler pass :class:`.WrapAngles` which is used to apply
+    angle constraints on gates in the :class:`.Target`. If a :class:`.Target` has defined
+    angle bounds this pass will analyze all the parameters for the gates in the circuit and
+    check that the target. For example, if a target contains a custom gate that has angle bounds
```
```suggestion
    check that the target conforms to the bounds. For example, if a target contains a custom gate that has angle bounds
```
I think this sentence was left incomplete.

mtreinish replied:
Updated for clarity in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

kt474 reviewed (commented):
Minor typos, the API and naming of classes/methods LGTM

raynelfss commented on crates/transpiler/src/passes/wrap_angles.rs:
```diff
@@ -0,0 +1,92 @@
+        let params: Vec<_> = inst
+            .params
+            .as_ref()
+            .unwrap()
```
Would you be able to use `.params_view()` for this??

mtreinish replied:
Probably, I'll give it a try locally.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

raynelfss commented on crates/transpiler/src/passes/wrap_angles.rs:
```diff
@@ -0,0 +1,92 @@
+                    &new_dag,
+                    None,
+                    None,
+                )?;
```
Would you be able to use the rust native `substitute_node_with_dag` that was added by #14766?

mtreinish replied:
Most likely. I wrote this long before that PR merged or was created so it wasn't an option then. But I suspect it's what I really need here.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339 and by extension no `py` is needed anymore.

raynelfss commented on crates/transpiler/src/target/bounds.rs:
```diff
@@ -0,0 +1,45 @@
+use smallvec::SmallVec;
+
+#[derive(Clone, Debug)]
+pub(crate) struct AngleBound(SmallVec<[Option<[f64; 2]>; 3]>);
```
Include a docstring here, to describe what this struct does.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

raynelfss commented on crates/transpiler/src/passes/gates_in_basis.rs:
```diff
@@ -39,6 +39,24 @@ pub fn gates_missing_from_target(dag: &DAGCircuit, target: &Target) -> PyResult<
         if !target.instruction_supported(gate.op.name(), &qargs_mapped) {
             return Ok(true);
         }
+        if target.has_angle_bounds()
+            && target.gate_has_angle_bounds(gate.op.name())
+            && !gate.is_parameterized()
+        {
+            let params: Vec<f64> = gate
+                .params
+                .as_ref()
+                .unwrap()
+                .iter()
+                .map(|x| {
+                    let Param::Float(val) = x else { unreachable!() };
+                    *val
+                })
+                .collect();
+            if !target.gate_supported_angle_bound(gate.op.name(), &params) {
+                return Ok(true);
+            }
+        }
```
Wouldn't this be included in a call to `instruction_supported` with the `check_bounds` argument to True? Unless that was only added to the Python counterpart.

mtreinish replied:
This was only done in Python, the rust version of the function didn't seem to check angles from what I remember.

raynelfss replied:
This is something that needs to be added. I've opened #14957 to keep track of this.

raynelfss commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -625,6 +650,23 @@ impl Target {
                             return Ok(false);
                         }
                     }
+                    if check_angle_bounds
+                        && self.has_angle_bounds()
+                        && parameters.iter().all(|x| matches!(x, Param::Float(_)))
+                    {
+                        let params: Vec<f64> = parameters
+                            .iter()
+                            .map(|x| {
+                                let Param::Float(val) = x else { unreachable!() };
+                                *val
+                            })
+                            .collect();
+                        if self.angle_bounds.contains_key(&operation_name)
+                            && !self.gate_supported_angle_bound(&operation_name, &params)
+                        {
+                            return Ok(false);
+                        }
+                    }
```
Is there a reason why these weren't included in the Rust native version of `instruction_supported()`?

mtreinish replied:
IIRC I did this here because the rust version didn't check or support changing angles at all.

raynelfss commented on crates/transpiler/src/angle_bound_registry.rs:
```diff
@@ -0,0 +1,149 @@
+
+/// Registry of Angle Wrapping function
+///
+///
```
Are you adding more to the docstring here?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

raynelfss commented on crates/transpiler/src/angle_bound_registry.rs:
```diff
@@ -0,0 +1,149 @@
+        &self,
+        name: &str,
+        angles: Vec<f64>,
+        qubits: Vec<u32>,
```
Why use `Vec<u32>` in the signature, I thought `PhysicalQubit` implements `FromPyObject` iirc.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339 I think I might have written this before we added the trait implementations. This PR has been around for a while.

raynelfss commented on crates/transpiler/src/angle_bound_registry.rs:
```diff
@@ -0,0 +1,149 @@
+    }
+}
+
+pub struct WrapAngleRegistry {
```
Add a docstring to explain what this does.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

raynelfss commented on crates/transpiler/src/angle_bound_registry.rs:
```diff
@@ -0,0 +1,149 @@
+        self.registry.insert(name, CallbackType::Native(callback));
+    }
+
+    /// Get a replacement circuit for
```
Finish your docstrings 🙃

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14406/commits/f84003a9af4b9b4e1c907477f7f7e41049d80339

raynelfss reviewed (commented):
Just did a quick pass through the code, nothing major to add, just a couple questions.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into wrap-angles-constraints

mtreinish committed:
Update for review comments

raynelfss commented on qiskit/transpiler/passes/utils/wrap_angles.py:
```diff
@@ -0,0 +1,88 @@
+    """Wrap angles outside the bound specified in the target.
+
+    This pass will check all the gates in the circuit and check if there are any gates outside the
+    bound specified in the target. If any gates outside the bound are identified the callback in
```
```suggestion
    bound specified in the target. If any gates outside the bound are identified, the callback in
```

raynelfss commented on releasenotes/notes/add-wrap-instructions-32335ea783203be9.yaml:
```diff
@@ -0,0 +1,111 @@
+                qubits: The physical qubit indices that this gate is operating on
+
+            Returns:
+                The DAGCircuit of the equivalent circuit
```
```suggestion
            """Callback function to wrap RZ gate angles

            Args:
                angles: The list of floating point parameter values for the instance of RZGate in
                    the circuit
                qubits: The physical qubit indices that this gate is operating on

            Returns:
                The DAGCircuit of the equivalent circuit"""
```

raynelfss commented on qiskit/transpiler/passes/utils/wrap_angles.py:
```diff
@@ -0,0 +1,88 @@
+       target = Target(num_qubits=1)
+       target.add_instruction(RZGate(param), angle_bounds=[(0, 0.5)])
+
+       def callback(angles, _qubits):
+           angle = angles[0]
+           if angle > 0:
+               number_of_gates = angle / 0.5
+           else:
+               number_of_gates = (6.28 - angle) / 0.5
+           dag = DAGCircuit()
+           dag.add_qubits([Qubit()])
+           for _ in range(int(number_of_gates)):
+               dag.apply_operation_back(RZGate(0.5), [dag.qubits[0]])
+           return dag
```
I know this is just an example but the math is a bit erroneous here. Due to a couple reasons:
- We should keep the remainder of the angle as it is within the bounds of the gate in the specified `Target`.
- We should first extract the remainder of the angle vs 2*pi (6.28) as anything that goes over that is a full rotation and is redundant.

Roughly it should look closer to this:
```py
       from decimal use Decimal

       def callback(angles, _qubits):
           angle = angles[0]
           # Use decimal for floating point modulus
           true_angle = float(Decimal(abs(angle)) % Decimal(6.28))
           if angle > 0:
               number_of_gates = true_angle / 0.5
           else:
               number_of_gates = (6.28 - true_angle) / 0.5
           dag = DAGCircuit()
           dag.add_qubits([Qubit()])
           for _ in range(int(number_of_gates)):
               dag.apply_operation_back(RZGate(0.5), [dag.qubits[0]])
           
           # Account for remainder
           remainder = float(Decimal(true_angle) % Decimal(0.5))
           if remainder:
               dag.apply_operation_back(RZGate(remainder), [dag.qubits[0]])
           return dag
```

Of course, double check this before adding it.
As I said before, this is just an example and doesn't have to be super accurate but I just had the thought, feel free to disregard if anything.

mtreinish replied:
The example wasn't intended to be a real suggestion it was just a quick sketch of how to use the interface. If you think we should do a more real example I can build that, we already have an example in the tests.

mtreinish replied:
I clarified the language that it's not a useful transformation but a simple example to show the interface in: https://github.com/Qiskit/qiskit/pull/14406/commits/46d3c0459242cd4f6d46b1a3dba4860fdfd3ef87

mtreinish committed:
Merge branch 'main' into wrap-angles-constraints

raynelfss commented on qiskit/transpiler/target.py:
```diff
@@ -159,6 +159,38 @@ class Target(BaseTarget):
+    :class:`.WrapAngles` pass is used to enforce the angle bounds, for this
+    to work you need to provide a function to the :class:`.WrapAngleRegistry`
+    used by the pass. You can see more details on this in:
+    :ref:`angle-bounds-on-gates`.
```
```suggestion
    You can check if any operations in the target have angle bounds set with,
    :meth:`.has_angle_bounds` and also if a specific name in the target has
    angle bounds set with :meth:`.gate_has_angle_bounds`. Whether a particular
    set of parameter values conforms to the angle bounds can be checked
    with :meth:`.add_owned_angle_bound`. In the preset pass managers the
    :class:`.WrapAngles` pass is used to enforce the angle bounds, for this
    to work you need to provide a function to the :class:`.WrapAngleRegistry`
    used by the pass. You can see more details on this in:
    :ref:`angle-bounds-on-gates`.
```

I don't think we've exposed `.add_owned_angle_bound()` to Python just yet. So maybe remove it from here.

raynelfss reviewed (commented):
This is about ready to merge, I have a couple of comments I found mostly in documentation and on the example included in the pass's docstring. Other than that this is looking good.

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

mtreinish committed:
Merge remote-tracking branch 'origin/main' into wrap-angles-constraints

mtreinish committed:
Mention doc example isn't a useful transformation

mtreinish committed:
Update qiskit/transpiler/target.py

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -159,6 +159,38 @@ class Target(BaseTarget):
+    :meth:`.has_angle_bounds` and also if a specific name in the target has
+    angle bounds set with :meth:`.gate_has_angle_bounds`. Whether a particular
+    set of parameter values conforms to the angle bounds can be checked
+    with :meth:`.add_owned_angle_bound`. In the preset pass managers the
```
```suggestion
    with :meth:`.supported_angle_bound`. In the preset pass managers the
```

mtreinish committed:
Update qiskit/transpiler/target.py
